oak-blob-plugins/src/main/java/org/apache/jackrabbit/oak/plugins/blob/AbstractSharedCachingDataStore.java

```diff
@@ -204,7 +203,13 @@ public DataRecord getRecordIfStored(DataIdentifier dataIdentifier)
                 return new FileCacheDataRecord(this, backend, dataIdentifier, rec.getLength(),
                     rec.getLastModified());
             } catch (Exception e) {
-                LOG.error("Error retrieving record [{}]", dataIdentifier, e);
+                if (this instanceof CompositeDataStoreAware &&
```

Comment:
The abstract class may not be an instance of `CompositeDataStoreAware` but the concrete class may be (e.g. `S3DataStore`).
The issue here is the `LOG.error()` call when the record is not found.  When you have only a single data store configured and you try to retrieve a record from the configured data store, it is an error if the record cannot be found.

When you are using a composite data store, each delegate may be looking for the record if the location if the record is unknown.  Thus it is possible that one delegate may look for the record and not find it, and then a second delegate looks for it and does find it.  Logging this as an error is misleading because it causes an admin to think the system is failing or the data store may be corrupt, which could be the case with a normal data store but is not necessarily the case with a composite data store.

So this change was made to address an actual issue I saw during testing with error messages showing up in the logs even though the system was working.

Reply:
Yes, that should be addressed by the encoding we have discussed in the past and is true for all the delegates as well not just the AbstractSharedCachingDataStore implementations for e.g. FileDataStore/OakFileDataStore would also throw these errors. We cannot mask this error as that would lead to debugging real problems.